lib/charms/glauth_k8s/v0/ldap.py

```diff
@@ -0,0 +1,325 @@
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+PYDEPS = ["dacite~=1.8.0"]
+
+DEFAULT_RELATION_NAME = "ldap"
+
+
+def leader_unit(func: Callable) -> Callable:
+    @wraps(func)
+    def wrapper(
+        obj: Union["LdapProvider", "LdapConsumer"], *args: Any, **kwargs: Any
+    ) -> Optional[Any]:
+        if not obj.unit.is_leader():
+            return None
+
+        return func(obj, *args, **kwargs)
+
+    return wrapper
+
+
+@leader_unit
+def _update_relation_app_databag(
+    ldap: Union["LdapProvider", "LdapConsumer"], relation: Relation, data: dict
+) -> None:
+    if relation is None:
+        return
+
+    relation.data[ldap.app].update(data)
+
+
+@dataclass(frozen=True)
+class LdapProviderData:
+    ldap_uri: str
+    base_dn: str
+    bind_dn: str
+    bind_password: str
+
+
+@dataclass(frozen=True)
+class LdapConsumerData:
+    app: str
+    model: str
+
+
+class LdapRequestedEvent(RelationEvent):
+    """An event emitted when the LDAP integration is built."""
+
+    @property
+    def data(self) -> Optional[LdapConsumerData]:
+        if not self.relation.data:
+            return None
+
+        relation_data = self.relation.data[self.relation.app]
+        return from_dict(data_class=LdapConsumerData, data=relation_data)
+
+
+class LdapProviderEvents(ObjectEvents):
+    ldap_requested = EventSource(LdapRequestedEvent)
+
+
+class LdapReadyEvent(RelationEvent):
+    """An event when the LDAP related information is ready."""
+
+
+class LdapUnavailableEvent(RelationEvent):
+    """An event when the LDAP integration is unavailable."""
+
+
+class LdapConsumerEvents(ObjectEvents):
+    ldap_ready = EventSource(LdapReadyEvent)
+    ldap_unavailable = EventSource(LdapUnavailableEvent)
+
+
+class LdapProvider(Object):
+    on = LdapProviderEvents()
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str = DEFAULT_RELATION_NAME,
+    ) -> None:
+        super().__init__(charm, relation_name)
+
+        self.charm = charm
+        self.app = charm.app
+        self.unit = charm.unit
+        self._relation_name = relation_name
+
+        self.framework.observe(
+            self.charm.on[self._relation_name].relation_changed,
+            self._on_relation_changed,
+        )
+
+    @leader_unit
+    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handle the event emitted when the consumer charm provides the
+        necessary data."""
+
+        self.on.ldap_requested.emit(event.relation)
+
+    def update_relation_app_data(
+        self, relation_id: int, data: LdapProviderData
+    ) -> None:
+        """An API for the provider charm to provide the LDAP related
+        information."""
+
+        relation = self.charm.model.get_relation(
+            self._relation_name, relation_id
+        )
+        _update_relation_app_databag(self.charm, relation, asdict(data))
+
+
+class LdapConsumer(Object):
```

Comment:
Let's use the Juju terminology:
```suggestion
class LdapRequirer(Object):
```